src/hotspot/share/jfr/recorder/stacktrace/jfrStackTraceRepository.cpp

```diff
@@ -196,9 +220,9 @@ traceid JfrStackTraceRepository::add_trace(const JfrStackTrace& stacktrace) {
 }
 
 // invariant is that the entry to be resolved actually exists in the table
-const JfrStackTrace* JfrStackTraceRepository::lookup(unsigned int hash, traceid id) const {
+const JfrStackTrace* JfrStackTraceRepository::lookup(unsigned int hash, traceid id) {
```

Comment:
Thanks, Jaroslav - yes, indeed.

Reply:
I also attempted to have the artefact tagging happen lazily, e.g. methods and classes, making it a part of the stack trace resolution process. There is a problem with that approach concerning class unloading because a referenced class can unload, but since it does not yet have the proper tag, the framework does not intercept it. I elaborated on setting an "interest" tag to handle this, but that approach opens additional challenges related to visibility and concurrency. Hard problem to solve; I was unable to do it (yet).

Reply:
Probably can live without lazy tagging in the first iteration - still it will be much better than the current state.
And then we can keep on adding incremental improvements based on cost-benefit weighing.